### PR TITLE
[DTensor] Prevent squeeze from redistributing with strict_view

### LINKED ISSUE #174136: [DTensor] squeeze_.dim updates spec but doesn't squeeze local tensor
When squeezing on a sharded dimension, `squeeze_.dim` on DTensor updates the DTensor's spec/shape but does NOT actually execute squeeze on the underlying local tensor, causing a metadata mismatch.

## Reproduction

```bash
torchrun --nproc_per_node=2 repro.py
```

```python
import torch
import torch.distributed as dist
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.tensor import distribute_tensor, Shard

def main():
    dist.init_process_group(backend='gloo')
    rank = dist.get_rank()
    mesh = init_device_mesh('cpu', (2,))

    x = torch.arange(4).reshape(1, 4).float()
    dt = distribute_tensor(x, mesh, [Shard(0)])

    print(f'[rank {rank}] BEFORE: DTensor shape={dt.shape}, local={dt._local_tensor.shape}')
    dt.squeeze_(0)
    print(f'[rank {rank}] AFTER:  DTensor shape={dt.shape}, local={dt._local_tensor.shape}')

    dist.destroy_process_group()

if __name__ == "__main__":
    main()
```

**Output:**
```
[rank 0] BEFORE: DTensor shape=torch.Size([1, 4]), local=torch.Size([1, 4])
[rank 0] AFTER:  DTensor shape=torch.Size([4]), local=torch.Size([1, 4])
```

DTensor claims shape `[4]` but local tensor is still `[1, 4]`.

## Root Cause

In `_dispatch.py`, `_dispatch_fast_path_python_tail` has special handling for `squeeze_.dim`:

```python
if op_call == aten.squeeze_.dim:
    args[0]._spec = output_spec
    return return_and_correct_aliasing(op_call, args, kwargs, args[0])
```

This updates the spec but the local tensor is never squeezed. Manual squeeze works:

```python
dt._local_tensor.squeeze_(0)  # Works - shape becomes [4]
```

But `dt.squeeze_(0)` leaves local tensor unchanged.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

### PR DESCRIPTION
**Pull request by stmcgovern**
Fixes https://github.com/pytorch/pytorch/issues/174136
When squeeze_.dim triggers redistribution (e.g. squeezing a globally-singleton sharded dim forces Shard(0) → Replicate), redistribute_local_args creates a new local tensor. The dispatch path updated `_spec` but never assigned the new tensor back to `_local_tensor`, leaving the DTensor in an inconsistent state where the spec claims shape [4] but the local tensor is still [1, 4].

Update: Prevent squeeze, as a view op, from doing a redistribution using the  `view_strict` keyword.

~This PR syncs `_spec` and `_local_tensor`. It uses the inplace_view tag, so other inplace view ops (unsqueeze_, t_, transpose_) won't hit this, when strategies are registered for them.~

The C++ fast path in dispatchDTensorOp explicitly skips inplace ops, so no C++ change is needed.

### CODE REVIEW

File: torch/distributed/tensor/_dispatch.py [hunk 1]
```diff
@@ -476,21 +476,20 @@ def _dispatch_fast_path_python_tail(
...
+                if torch.Tag.inplace_view in op_call.tags:
+                    # Inplace view ops may change tensor meta. Update both
+                    # spec and local tensor -- redistribution may produce a
+                    # new local tensor.
+                    # See https://github.com/pytorch/pytorch/issues/174136
                     args[0]._spec = output_spec
-                    # use return_and_correct_aliasing to match the outer and the inner
-                    # aliasing. See https://github.com/pytorch/pytorch/pull/158954
+                    if not isinstance(local_results, torch.Tensor):
+                        raise AssertionError
```
Comment (wconstab) [inline torch/distributed/tensor/_dispatch.py]: nit: add text about what type the local result is

File: torch/distributed/tensor/_dispatch.py [hunk 2]
```diff
@@ -476,21 +476,20 @@ def _dispatch_fast_path_python_tail(
...
 
-                # NOTE: aten.squeeze_.dim is an inplace op but it also may change
-                # the inplace argument's tensor meta. Here we choose to special case
-                # this op because as far as I know this is the only inplace op that
-                # has such as behavior. We can extend this special case if necessary.
-                if op_call == aten.squeeze_.dim:
-                    # update the spec to handle tensor meta changes
+                if torch.Tag.inplace_view in op_call.tags:
+                    # Inplace view ops may change tensor meta. Update both
+                    # spec and local tensor -- redistribution may produce a
```
Comment (wconstab) [inline torch/distributed/tensor/_dispatch.py]: nit: comment says 'redistribution may produce..' but i think in this path a redistribution is actually not happening- its purely a metadata change right?

File: test/distributed/tensor/test_view_ops.py [hunk 3]
```diff
@@ -692,6 +692,33 @@ def test_squeeze_(self):
...
+        """
+        mesh = init_device_mesh(self.device_type, (self.world_size,))
+
+        # Squeeze a globally-singleton sharded dim: requires redistribution
+        # (Shard(0) -> Replicate) since the sharded dim disappears.
+        x = torch.arange(4, dtype=torch.float32, device=self.device_type).reshape(1, 4)
+        dt = distribute_tensor(x, mesh, [Shard(0)])
+        dt.squeeze_(0)
+        self.assertEqual(dt.shape, torch.Size([4]))
+        self.assertEqual(dt._local_tensor.shape, torch.Size([4]))
```
Comment (wconstab) [inline test/distributed/tensor/test_view_ops.py]: now i'm wondering- is this squeeze actually doing a redistribution? or just changing placements (aka view operation)?  lets add comm counter to disambiguate this in the test. (and update the comment above if needed)

File: test/distributed/tensor/test_view_ops.py [hunk 4]
```diff
@@ -692,6 +692,41 @@ def test_squeeze_(self):
...
+        """
+        mesh = init_device_mesh(self.device_type, (self.world_size,))
+        c10d_fn = torch.ops.c10d_functional
+
+        # Shard(0) on a size-1 dim → Replicate via allgather.
+        x = torch.arange(4, dtype=torch.float32, device=self.device_type).reshape(1, 4)
+        dt = distribute_tensor(x, mesh, [Shard(0)])
+        comm_mode = CommDebugMode()
+        with comm_mode:
+            dt.squeeze_(0)
```
Comment (wconstab) [inline test/distributed/tensor/test_view_ops.py]: related to my original comment, it seems like this PR is going against the design for dtensor views. Views are supposed to be gauranteed fast and sharing same memory as the original tensor, so doing a redistribution implicitly during a view operation is semantics breaking even if it's convenient in some cases.  We prefer to error as unsupported in such cases and require the user intervene to do a redistribution first manually and then perform a true view operation

Comment (stmcgovern) [reply test/distributed/tensor/test_view_ops.py]: I see. Yes this PR is cleaning up after a redistribution. But that redistribution should be banned. I'll use strict_view=True for squeeze, following https://github.com/pytorch/pytorch/pull/149764

**Comment by wconstab** [comment]
@claude review the PR

File: torch/distributed/tensor/_dispatch.py [hunk 5]
```diff
@@ -496,29 +491,37 @@ def _dispatch_fast_path_python_tail(
...
-                    # Inplace operations that change placement are not supported because
-                    # they would require redistribution, which breaks aliasing semantics.
-                    # If there are views into the tensor, the views would not be updated.
-                    if args[0]._spec.placements != output_spec.placements:
+                # Check if arg 0's input placement actually needs to change.
+                # needs_redistribute can be True for reasons other than
+                # arg 0 placement changes (e.g. squeeze dim rewriting),
+                # so we check the redistribute schema directly.
+                if (
+                    output_sharding.needs_redistribute
```
Comment (wconstab) [inline torch/distributed/tensor/_dispatch.py]: do we need this whole codepath still, after using strict_view in the right places? what are the expected cases where we'd end up in this block?

Comment (stmcgovern) [reply torch/distributed/tensor/_dispatch.py]: I think you're right that it's overly defensive.

Comment (zpcore) [reply torch/distributed/tensor/_dispatch.py]: Can you remove this checking? Otherwise lgtm

Comment (stmcgovern) [reply torch/distributed/tensor/_dispatch.py]: thanks- will do

Comment (stmcgovern) [reply torch/distributed/tensor/_dispatch.py]: @zpcore I pushed the removal.

**Review by zpcore — APPROVED** [review]

File: torch/distributed/tensor/_dispatch.py (hunk 5 above)
Comment (zpcore) [reply torch/distributed/tensor/_dispatch.py]: Thanks for the fixing, LGTM!

**Comment by stmcgovern** [comment]
these failures seem unrelated. going to try to merge
